### PR TITLE
intercom stuff

### PR DESCRIPTION
**Pull request by Flask**

### CODE REVIEW

File: src/Classes/Product/KoalaFarm.php [hunk 1]
```diff
@@ -104,4 +107,33 @@ public function notificationMessage(string $event, array $params): array
...
+    }
+
+    public static function IntercomEvent(string $email, string $event, array $metadata): bool
+    {
+        $apiKey = ServerUtility::get("INTERCOM_API_KEY");
+        if (!$apiKey) {
+            return false;
+        }
+
+        $client = new IntercomClient($apiKey);
```
Comment (mweibel) [inline src/Classes/Product/KoalaFarm.php]: hrm.. a client per call sucks a bit. But it's PHP and we have probably only one event per request anyway soo.. it doesn't really matter?

File: src/Classes/Product/KoalaFarm.php [hunk 2]
```diff
@@ -104,4 +107,33 @@ public function notificationMessage(string $event, array $params): array
...
+
+    public function notify($email, $event, $params): void
+    {
+        self::IntercomEvent($email, $event, $params);
+    }
+
+    public static function IntercomEvent(string $email, string $event, array $metadata): bool
+    {
+        $apiKey = ServerUtility::get("INTERCOM_API_KEY");
+        if (!$apiKey) {
```
Comment (mweibel) [inline src/Classes/Product/KoalaFarm.php]: `ServerUtility::get("INTERCOM_API_KEY");` will throw if the default is not changed from `null` (second param)

File: src/Classes/Product/KoalaFarm.php [hunk 3]
```diff
@@ -104,4 +107,33 @@ public function notificationMessage(string $event, array $params): array
...
+            $client->events->create([
+                "email" => $email,
+                "event_name" => $event,
+                "created_at" => time(),
+                "metadata" => $metadata
+            ]);
+
+            return true;
+        } catch (\Exception $e) {
+            LogHelper::warn("Could not send Event to Intercom: " . $e->getMessage());
```
Comment (mweibel) [inline src/Classes/Product/KoalaFarm.php]:
```suggestion
            LogHelper::warn("could not send Event to Intercom", ["err": $e, "event" => $event, "email" => $email, "metadata" => $metadata]);
```

File: src/Classes/Product/KoalaFarm.php [hunk 4]
```diff
@@ -104,4 +107,33 @@ public function notificationMessage(string $event, array $params): array
...
+        if (!$apiKey) {
+            return false;
+        }
+
+        $client = new IntercomClient($apiKey);
+        try {
+            $client->events->create([
+                "email" => $email,
+                "event_name" => $event,
+                "created_at" => time(),
```
Comment (mweibel) [inline src/Classes/Product/KoalaFarm.php]: Hmm. they just accept a unix timestamp?
What about timezones?

File: src/Classes/Product/Product.php [hunk 5]
```diff
@@ -23,10 +23,12 @@ public function confirmationMailContent(): array;
     public function notificationMailTemplate(): array;
 
     /**
-     * @param  string                    $event
-     * @param  array                     $params
+     * @param string $event
+     * @param array $params
```
Comment (mweibel) [inline src/Classes/Product/Product.php]: please run `./vendor/bin/php-cs-fixer f` if you have PHP locally, or if not: `docker-compose exec panel ./vendor/bin/php-cs-fixer f`

File: src/Classes/Utility/NotificationUtility.php [hunk 6]
```diff
@@ -92,7 +96,7 @@ public static function alertAdmin(string $content = ''): bool
         }
     }
 
-    protected static function sendMail(string $recipient, string $subject, $content, array $button, array $from = ['hello@idling.host' => 'Helio'], string $htmlTemplate = null): bool
+    public static function sendMail(string $recipient, string $subject, $content, array $button, array $from = ['hello@idling.host' => 'Helio'], string $htmlTemplate = null): bool
```
Comment (mweibel) [inline src/Classes/Utility/NotificationUtility.php]:
```suggestion
    protected static function sendMail(string $recipient, string $subject, $content, array $button, array $from = ['hello@idling.host' => 'Helio'], string $htmlTemplate = null): bool
```

File: src/Classes/Utility/NotificationUtility.php [hunk 7]
```diff
@@ -36,15 +36,19 @@ public static function sendConfirmationMail(User $user, Product $product, string
...
 
+        $product->notify($user->getEmail(), "userRegistration", []);
+
         return static::sendMail($user->getEmail(), $subject, $content, ['text' => 'Confirm Email & Login', 'link' => $confirmLink], $product->emailSender(), $product->emailHTMLLayout());
     }
 
-    public static function notifyUser(User $user, Product $product, string $templateName, array $params): bool
+    public static function notifyUser(User $user, Product $product, string $event, array $params): bool
     {
+        $product->notify($email, $event, $params);
```
Comment (mweibel) [inline src/Classes/Utility/NotificationUtility.php]:
```suggestion
        $product->notify($user->getEmail(), $event, $params);
```

File: .panel.env.dist [hunk 8]
```diff
@@ -59,3 +59,6 @@ BLENDER_STORAGE_CREDENTIALS_JSON_PATH=/var/www/cnf/google-auth-blender-storage.j
 KOALA_FARM_ORIGIN=http://localhost:3000
 
 BASE_URL=http://localhost
+
+# Intercom API KEY
+# INTERCOM_API_KEY=REPLACE_WITH_CORRECT_VALUE
```
Comment (mweibel) [inline .panel.env.dist]: 😂 
please replace the comment with a valuable comment or remove it 😉

File: src/Classes/Utility/NotificationUtility.php [hunk 9]
```diff
@@ -36,15 +36,19 @@ public static function sendConfirmationMail(User $user, Product $product, string
...
+
         try {
-            $template = $product->notificationMessage($templateName, $params);
+            $template = $product->notificationMessage($event, $params);
         } catch (\InvalidArgumentException $e) {
-            LogHelper::info("template name ${templateName} not supported for product", ['product title' => $product->title()]);
+            LogHelper::info("template name ${$event} not supported for product", ['product title' => $product->title()]);
```
Comment (mweibel) [inline src/Classes/Utility/NotificationUtility.php]:
```suggestion
            LogHelper::info("template name ${event} not supported for product", ['product title' => $product->title()]);
```

**Comment by mweibel** [comment]
```./vendor/bin/phpunit --stop-on-failure --stop-on-error```

😂 
(I'm guilty of this as well, many times over. Right @cbuchli?) 😄

File: src/Classes/Product/KoalaFarm.php (hunk 4 above)
Comment (Flask) [reply src/Classes/Product/KoalaFarm.php]: yep. only timestamp is accepted according to the api specs

Comment (cbuchli) [reply src/Classes/Product/KoalaFarm.php]: According to [the Docs](https://developers.intercom.com/intercom-api-reference/reference#submitting-events) it's UTC. So if this time is needed anywhere, we should submit the proper one...

Quickest would be:
```
(new \DateTime('now', ServerUtility::getTimezoneObject() ))->setTimezone(new \DateTimeZone(\DateTimeZone::UTC))->getTimestamp()```

Maybe put it in ServerUtility for re-use...

File: src/Classes/Product/KoalaFarm.php [hunk 10]
```diff
@@ -104,4 +106,33 @@ public function notificationMessage(string $event, array $params): array
...
+        try {
+            $client->events->create([
+                'email' => $email,
+                'event_name' => $event,
+                'created_at' => ServerUtility::getCurrentUTCTimestamp(),
+                'metadata' => $metadata,
+            ]);
+
+            return true;
+        } catch (\Exception $e) {
```
Comment (mweibel) [inline src/Classes/Product/KoalaFarm.php]: actually, could you catch  `\Throwable $t` instead? Just to ensure nothing can ever crash the request.

File: src/Classes/Utility/MiddlewareForHttpUtility.php [hunk 11]
```diff
@@ -135,7 +135,7 @@ public static function addMiddleware(App $app): void
         $app->add(new CorsMiddleware([
             'logger' => LogHelper::getInstance('cors'),
             'origin' => ['*'],
-            'methods' => ["GET", "POST", "PUT", "PATCH", "DELETE", "HEAD"],
+            'methods' => ['GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'HEAD'],
```
Comment (mweibel) [inline src/Classes/Utility/MiddlewareForHttpUtility.php]: oops. thanks for fixing :)

**Pull request merged**